### ISSUE TITLE
Angle parameter in motion_blur not updating during optimization

### ISSUE THREAD

**Issue by alex-flwls** [kornia/kornia#1993]
### Describe the bug

I want to match the level of motion blur between two images, so I am using Kornia's optimization ability to optimize the parameters of the kornia.filters.motion_blur function. Specifically, kernel_size, angle, and direction. The direction parameter optimizes fine (the parameter changes over time and it has gradients), but the angle parameter does not. It does not update or change over time, and the gradients are zero.

My code is heavily based on the Image Alignment by Homography Optimization example in the tutorials (https://kornia-tutorials.readthedocs.io/en/latest/homography.html). 

### Reproduction steps

```bash
1. Run this code and inspect the angle tensor and its gradients: 

# import libraries
from typing import List
import os

import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import kornia
# from kornia.filters.kernels_geometry import get_motion_kernel2D
# from kornia.filters.filter import filter2d
import cv2

import sys, os
sys.path.insert(0, os.path.abspath('..')) 
import utils.data_io as io


# hyperparameters
learning_rate: float = 1e-4  # the gradient optimisation update step
num_iterations: int = 100  # the number of iterations until convergence
num_levels: int = 6  # the total number of image pyramid levels
error_tol: float = 1e-8  # the optimisation error tolerance

log_interval: int = 100  # print log every N iterations
use_cuda: bool = torch.cuda.is_available()
device = torch.device('cuda' if use_cuda else 'cpu')
print('Using ', device)


def load_image(file_name: str):
    '''
    Loads the image with OpenCV and converts to torch.Tensor                                      
    '''
    assert os.path.isfile(file_name), "Invalid file {}".format(file_name)

    # load image                                                                        
    img: np.ndarray = io.exr_to_numpy(file_name)

    # clip image and normalise to 0 - 1
    img = img / np.amax(img)
    img = np.clip(img, 0, 1)

    # convert image to torch tensor                                                                  
    tensor: torch.Tensor = kornia.utils.image_to_tensor(img).float() # CxHxW
    
    return tensor[None] # 1xCxHxW


class BlurParams(nn.Module):
    def __init__(self, init_value):
        super(BlurParams, self).__init__()
        self.blur_params = nn.Parameter(torch.tensor(init_value), requires_grad=True)
        # self.reset_parameters()

    def reset_parameters(self):
        torch.nn.init.eye_(self.blur_params)

    def forward(self):
        return torch.unsqueeze(self.blur_params, dim=0)  # 1x1x1 


def get_gaussian_pyramid(img: torch.Tensor, num_levels: int):
    """Utility function to compute a gaussian pyramid."""
    pyramid = []
    pyramid.append(img)
    for _ in range(num_levels - 1):
        img_curr = pyramid[-1]
        img_down = kornia.geometry.pyrdown(img_curr)
        pyramid.append(img_down)
    return pyramid


def compute_scale_loss(img_src: torch.Tensor,
                       img_dst: torch.Tensor,
                       kernel: nn.Module,
                       angle,
                       direction,
                       optimizer: torch.optim,
                       num_iterations: int,
                       error_tol: float,
                       kernel_size):
    assert len(img_src.shape) == len(img_dst.shape), (img_src.shape, img_dst.shape)
    
    # init loop parameters
    loss_tol = torch.tensor(error_tol)
    loss_prev = torch.finfo(img_src.dtype).max
    
    for i in range(num_iterations):
        # create blur params
        # kernel_inv: torch.Tensor = torch.inverse(kernel())
        # if angle.shape[0] == 1:
        #     angle_inv = angle
        # else:
        #     angle_inv: torch.Tensor = torch.inverse(angle)
        # direction_inv: torch.Tensor = torch.inverse(direction())
        # angle -= torch.tensor(3.14/32)
        angle.retain_grad()
        direction.retain_grad()

        img_src_to_dst = kornia.filters.motion_blur(img_src, kernel_size=kernel_size, angle=angle, direction=direction, mode='nearest', border_type='circular')
        # img_src_to_dst = img_src * angle * direction

        print("angle: ", angle, "direction: ", direction)
        

        # compute and mask loss
        loss = F.l1_loss(img_src_to_dst, img_dst, reduction='none') 

        # ones = kornia.filters.motion_blur(torch.ones_like(img_src), kernel_size, angle, direction)
        # loss = loss.masked_select((ones > 0.9)).mean()
        # loss = loss * 10
        loss = loss.mean() 

        print("loss:, ", loss.data)

        # compute gradient and update optimizer parameters
        optimizer.zero_grad()
        loss.backward()
        # angle.grad += torch.tensor(0.0001)
        optimizer.step()


### run the optimisation ###

img_src: torch.Tensor = load_image('/image_path').to(device)
img_dst: torch.Tensor = load_image('/image_path').to(device)

height, width = img_src.shape[-2:]

kernel_size = 13

# create blur parameters
kernel = BlurParams(3.0).to(device)
angle = BlurParams(1.6).to(device)
direction = BlurParams(0.1).to(device)

# create optimizer
optimizer = optim.Adam([*angle.parameters(), *direction.parameters()], lr=learning_rate)

# send data to device
img_src, img_dst = img_src.to(device), img_dst.to(device)

# compute the gaussian pyramids
img_src_pyr: List[torch.Tensor] = get_gaussian_pyramid(img_src, num_levels)
img_dst_pyr: List[torch.Tensor] = get_gaussian_pyramid(img_dst, num_levels)

# pyramid loop

for iter_idx in range(num_levels):
    # get current pyramid data
    scale: int = (num_levels - 1) - iter_idx
    img_src = img_src_pyr[scale]
    img_dst = img_dst_pyr[scale]

    # compute scale loss
    compute_scale_loss(img_src, img_dst, kernel, angle(), direction(),
                       optimizer, num_iterations, error_tol, kernel_size=kernel_size)

    print('Optimization iteration: {}/{}'.format(iter_idx+1, num_levels))
       
    # merge warped and target image for visualization
    img_src_to_dst = kornia.filters.motion_blur(img_src, kernel_size, angle(), direction())
    
    frame = img_src_to_dst.detach().numpy()
    new_frame = np.moveaxis(frame[0]*255, 0, -1)
    new_frame = cv2.cvtColor(new_frame, cv2.COLOR_BGR2RGB)

    # show data
    # io.save_image_as_exr(frame, f"/image_path")    
    cv2.imwrite(f"/image_path", new_frame)
```

### Expected behavior

For the angle tensor to update over time, like the direction tensor.

angle:  tensor([1.6000], grad_fn=<AsStridedBackward0>) direction:  tensor([0.0829], grad_fn=<AsStridedBackward0>)
angle:  tensor([1.6000], grad_fn=<AsStridedBackward0>) direction:  tensor([0.0824], grad_fn=<AsStridedBackward0>)
angle:  tensor([1.6000], grad_fn=<AsStridedBackward0>) direction:  tensor([0.0820], grad_fn=<AsStridedBackward0>)

When printed out, gradients for angle are always zero. 

### Environment

```shell
Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.29.3)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:38:29) [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.2
[pip3] torch==1.12.1
[conda] numpy                     1.23.2          py310h127c7cf_0    conda-forge
[conda] torch                     1.12.1                   pypi_0    pypi
```

**Comment by edgarriba** [comment]
have you tried to run this piece of code with autograd anomaly detection?
https://pytorch.org/docs/stable/autograd.html#anomaly-detection

otherwise this chained function needs deep debugging /cc @shijianjian

**Comment by alex-flwls** [comment]
Thanks for your reply. I just tried the autograd anomaly detection, but because the code does not error out, the gradients are just zero and the parameter does not update (i.e. it remains constant), the anomaly detection is not used.

I agree there is some deep debugging needed, I have sunk a lot of time into this trying to debug it. I appreciate your time either way.

**Comment by edgarriba** [comment]
@alex-flwls feel free to join our slack for further discussing in public channels

**Comment by edgarriba** [comment]
If anyone is interested in this, feel free to re-open the ticket. Thanks!
